jewels.py

```diff
@@ -0,0 +1,278 @@
+import math
+import re
+from queue import Queue
+import data
+
+legion_passive_effects = data.legion_passive_mapping()
+notable_hashes_for_jewels = [
+    '26725', '36634', '33989', '41263', '60735', '61834', '31683', '28475', '6230', '48768', '34483', '7960',
+    '46882', '55190', '61419', '2491', '54127', '32763', '26196', '33631', '21984', '29712', '48679', '9408',
+    '12613', '16218', '2311', '22994', '40400', '46393', '61305', '12161', '3109', '49080', '17219', '44169',
+    '24970', '36931', '14993', '10532', '23756', '46519', '23984', '51198', '61666', '6910', '49684', '33753',
+    '18436', '11150', '22748', '64583', '61288', '13170', '9797', '41876', '59585'
+]
+
+
+class TreeGraph:
+    """This class is used to determine the shortest paths between two passive skills in a given passive tree"""
+    def __init__(self, tree, skills):
+        # skills['hashes'] can contain hashes that are not part of the tree (cluster notables
+        self.node_hashes = {str(hash) for hash in skills['hashes']} & \
+                           set(node_hash for node_hash in tree['nodes'] if node_hash != 'root')
+        self.tree = tree
+        self.id_for_hash = {node: idx for idx, node in enumerate(self.node_hashes)}
+        self.adjacency_list = {node: set() for node in range(len(self.node_hashes))}
+        self.fill_adjacency_list()
+
+    def fill_adjacency_list(self):
+        for idx, node_hash in enumerate(self.node_hashes):
+            node = self.tree['nodes'].get(node_hash)
+            if node is None:
+                continue
+            for neighbour_hash in ((set(node.get('in', [])) | set(node.get('out', []))) & self.node_hashes):
+                self.add_edge(idx, self.id_for_hash[neighbour_hash])
+
+    def add_edge(self, node1, node2):
+        self.adjacency_list[node1].add(node2)
+        self.adjacency_list[node2].add(node1)
+
+    def bfs(self, start_node_hash, target_node_hash):
+        start_node = self.id_for_hash[start_node_hash]
+        target_node = self.id_for_hash[target_node_hash]
+        queue = Queue()
```

Comment:
`queue.Queue` is synchronized (read: for multithreaded applications). probably use https://docs.python.org/3/library/collections.html#collections.deque instead